Add IME support for Windows

lunabunn opened the pull request:
See https://github.com/kettle11/kapp/issues/64 for additional details

kettle11 commented on kapp_platform_common/src/events.rs:
```diff
@@ -33,6 +33,16 @@ pub enum Event {
     CharacterReceived {
         character: char,
     },
+    /// The composition status of the IME.
```
This is a nitpick, but generally (when I remember) I like to define acronyms the first time they appear in code or in a document.

lunabunn replied:
But considering these are doc comments, wouldn't it be a bit weird to document the acronym for `IMEComposition` but not for `IMEEndComposition`?

kettle11 commented on kapp_platforms/src/windows/event_loop_windows.rs:
```diff
@@ -36,6 +36,41 @@ pub unsafe extern "system" fn window_callback(
+            produce_event(Event::CharacterReceived { character });
+        }
+        WM_IME_STARTCOMPOSITION => {
+            return TRUE as isize;
```
Are these `return TRUE as isize;` lines necessary for the composition events? 

I see no mention in the win32 docs about an expected return value.

lunabunn replied:
Actually, you're right. `return 0` might suffice there. However, do note that there are also no mentions of an expected return value for WM_IME_COMPOSITION, yet example code I found online seems to distinguish return values of TRUE vs. 0.

kettle11 replied:
I was actually thinking maybe no return at all is needed here. Is that not the case?

lunabunn commented:
Note that this only implements IME *events*. Functionality to set the location of IME candidate lists is not yet exposed (mainly because I myself don't have a usecase for it).

kettle11 commented on kapp_platforms/src/windows/event_loop_windows.rs:
```diff
@@ -36,6 +36,41 @@ pub unsafe extern "system" fn window_callback(
+            if l_param as u32 & GCS_COMPSTR != 0 {
+                let len = ImmGetCompositionStringW(himc, GCS_COMPSTR, null_mut(), 0);
+                if len == 0 {
+                    return 0;
```
`ImmReleaseContext(hwnd, himc);` will not be called in this case, which seems like a small bug.

Also should `IMEComposition` events be sent if IMEComposition is empty?

lunabunn replied:
Nice catch! I don't think it should be sent if it's empty, although I'm not sure what the generally accepted behaviour would be.

kettle11 replied:
This new version still has the same bug if `size_bytes` == 0

It seems like `ImmReleaseContext(hwnd, himc);` could be at the end of the `WM_IME_COMPOSITION` block.

Also `buffer.set_len(size_bytes as usize / 2);` should b changed to match `  let mut buffer = Vec::<u16>::with_capacity(size_bytes as usize / std::mem::size_of::<u16>());`

lunabunn commented:
Changed all return values to 0 as I couldn't find any documented differences. Also added `(Input Method Editor)` to the docs for `Event::IMEComposition`. One wouldn't really be using `Event::IMEEndComposition` without first using `Event::IMEComposition`, so I think that should be fine.

kettle11 commented on kapp_platforms/src/windows/event_loop_windows.rs:
```diff
@@ -36,6 +36,41 @@ pub unsafe extern "system" fn window_callback(
+                return 0;
+            }
+            if l_param as u32 & GCS_COMPSTR != 0 {
+                let len = ImmGetCompositionStringW(himc, GCS_COMPSTR, null_mut(), 0);
```
I'd prefer `size_bytes` or something similar instead of `len` here. Because `len` is ambiguous if it's indicating the number of characters or the size of the buffer.

kettle11 commented on kapp_platforms/src/windows/event_loop_windows.rs:
```diff
@@ -36,6 +36,41 @@ pub unsafe extern "system" fn window_callback(
+                if len == 0 {
+                    return 0;
+                }
+                let mut buffer = Vec::<u16>::with_capacity(len as usize / 2);
```
This is a nitpick, but I'd prefer `/ std::mem::size_of::<u16>()` here because it makes more explicit what's going on.

kettle11 commented on kapp_platforms/src/windows/event_loop_windows.rs:
```diff
@@ -36,6 +36,41 @@ pub unsafe extern "system" fn window_callback(
+                }
+                let mut buffer = Vec::<u16>::with_capacity(len as usize / 2);
+                ImmGetCompositionStringW(himc, GCS_COMPSTR, buffer.as_mut_ptr().cast(), len as u32);
+                buffer.set_len(len as usize / 2);
```
I didn't know about `set_len` :O 

That could be very handy.

lunabunn commented:
@kettle11 Sorry, I guess I'm way too sleepy right now to code properly😅 
Everything should be good now, this time for real! (at least I hope)

kettle11 commented:
I can relate. Half of `kapp`'s comments are so I can partially understand later when I'm sleepy. Get some rest!

This looks great though! I'll merge this now.

(Thought I still think there might be a tiny bug with the `ImmReleaseContext` spot, but we can catch it in a follow-up)
